From 18196062e212ddbc0bcc0be9089b78022ae9cb6c Mon Sep 17 00:00:00 2001
From: Sanyog Kale <skale@habana.ai>
Date: Thu, 05 May 2022 13:07:37 +0300
Subject: [PATCH] [SW-0] gaudi3: add manf_mode gpio check for RMA

It is confirmed by system team that GPIO_22 (MANF_MODE) needs to be checked
for 0 value along with module_id[4..3] from strap to enter RMA mode.
This patch updates RMA condition to check for MANF_MODE.

Change-Id: Ifa824795d0f88088e88baf5256c505afbf852494
---

diff --git a/arch/gaudi3/arc/common.c b/arch/gaudi3/arc/common.c
index e326f31..d1b6cbf 100644
--- a/arch/gaudi3/arc/common.c
+++ b/arch/gaudi3/arc/common.c
@@ -367,10 +367,12 @@
 }
 
 #define BOOT_STRAP_RMA_MODE_MASK			0x3
+#define GAUDI_MANF_MODE_GPIO				22
 
 void arcboot_rom_fuse_override(void)
 {
-	uint32_t rma_mode_status;
+	uint32_t rma_mode_status, rma_mode_gpio;
+	uint32_t index = GAUDI_MANF_MODE_GPIO % GPIOS_PER_BANK;
 	uint32_t strap_h = regmap_readl(mmD0_PSOC_BOOT_CONF_BOOT_STRAP_PINS_H);
 
 	if (arcboot_rom_global.config_fuse & EFUSE_SYSCFG_EOM_MASK)
@@ -380,10 +382,13 @@
 			>> D0_PSOC_BOOT_CONF_BOOT_STRAP_PINS_H_MODULE_ID_S;
 
 	rma_mode_status >>= BOOT_STRAP_RMA_MODE_MASK;
-	/* TODO: add a relavent gpio check for RMA mode once finalized for gaudi3 */
-	if ((rma_mode_status & BOOT_STRAP_RMA_MODE_MASK) == BOOT_STRAP_RMA_MODE_MASK)
+	rma_mode_gpio = regmap_readl(mmD0_PSOC_GPIO1_DATA) & BIT(index);
+
+	if (((rma_mode_status & BOOT_STRAP_RMA_MODE_MASK) == BOOT_STRAP_RMA_MODE_MASK) &&
+		(rma_mode_gpio == 0)) {
 		/* Reset the efuse pre-production bits to 0(pre-production value) */
 		arcboot_rom_global.config_fuse &= ~EFUSE_SYSCFG_PRE_PRODUCTION_BITS;
+	}
 
 	/* KEY index 4 must not be revoked */
 	arcboot_rom_global.config_fuse &= ~EFUSE_SYSCFG_REVOKE_KEY_4;
